Aggregator account

gregorylegarec opened the pull request:
This PR handles konnector declaring `agregator` field in their manifest.

An aggregator is a main service used by several konnectors (for example:
Budget Insight for Banking konnectors).

When an account is created, we check if the konnector declares an aggregator.
If so, we create a _master account_, which will be later filled by the konnector.

Then we relate every account of the given konnector to this master account.

This PR use CozyClient but still mix some old logic from redux-cozy-client
to keep the existant behaviour working. It also adds realtime for accounts
to keep the whole state synchronized.

gregorylegarec commented:
@ptbrowne Honestly, it will be a lot of work. I can maybe add test to `connections/accounts` but it is almost testing that cozy-client is called as expected.

CPatchane commented on src/containers/AccountConnection.jsx:
```diff
@@ -267,13 +275,25 @@ class AccountConnection extends Component {
 
+    const newAccount = {
+      account_type: konnector.slug,
+      accountType: konnector.slug,
```
Why twice the information here?

sizvix replied:
new name with camel case and old name for compatibility ?

gregorylegarec replied:
After a little thinking we should have only one prop at a time, so let's keep account_type for now.

CPatchane commented on src/lib/CollectStore.js:
```diff
@@ -127,10 +137,9 @@ export default class CollectStore {
     const connection = {
+      account: account,
       konnector: konnector
     }
```
Just that just be enough here
```js
const connection = { account, konnector }
```

CPatchane commented on src/connections/accounts.js:
```diff
@@ -0,0 +1,66 @@
+
+  const createMasterAccount = async id => {
+    const { data } = await client.create(ACCOUNTS_DOCTYPE, {
+      _id: id
```
Is it safe or what we want to force the `_id` here?

sizvix commented on src/connections/accounts.js:
```diff
@@ -0,0 +1,66 @@
+
+  const createSlaveAccount = async (konnector, attributes) => {
+    const masterAccountId =
+      konnector.aggregator && konnector.aggregator.accountId
```
This line supposed `masterAccountId` could be undefined.
Is it not a problem for `values` l.46
and `_id` l.55 ?

Crash-- commented on src/connections/accounts.js:
```diff
@@ -0,0 +1,66 @@
+        _id: id
+      })
+    )
+    return data[0]
```
je m'incruste vite fait :D 

tu dois pouvoir faire un ```
const {data } = await client.query(client.get(ACCOUNTS_DOCTYPE, id));
return data
```

gregorylegarec replied:
@Crash-- Le comportement est finalement un poil différent, `get` retourne une 404 avec les inconvénients que cela implique, où `find.where` renvoie une liste vide, ce qui m'arrange particulièrement bien.

gregorylegarec replied:
@Crash-- Update : c'est en fait dû à un problème dans le code, voir cette issue https://github.com/cozy/cozy-client/issues/301

gregorylegarec replied:
En gros une 404 génère ensuite inévitablement une typeError

gregorylegarec commented:
Ok I added some tests with 100% coverage on accounts connection (it makes me throw exception, so yes, it was a good thing).

Also, I split the original commit into two, to differenciate the accounts mutations part (which will stay in cozy-harvest) and the "making the thing work in Cozy-Home" part (which will disappear soon).

gregorylegarec commented on src/targets/browser/index.jsx:
```diff
@@ -58,7 +60,21 @@ document.addEventListener('DOMContentLoaded', () => {
+        attributes: {},
+        relationships: {
+          master: {
+            type: 'has-one',
```
:warning: Waiting for https://github.com/cozy/cozy-client/pull/304
